src/Microsoft.AspNetCore.Server.Kestrel/KestrelWebHostBuilderExtensions.cs

```diff
@@ -1,16 +1,31 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Reflection;
+using Microsoft.AspNetCore.Hosting.Server;
 using Microsoft.AspNetCore.Server.Kestrel;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
 
 namespace Microsoft.AspNetCore.Hosting
 {
     public static class KestrelWebHostBuilderExtensions
     {
         public static IWebHostBuilder UseKestrel(this IWebHostBuilder hostBuilder)
         {
-            return hostBuilder.UseServer(typeof(KestrelServer).GetTypeInfo().Assembly.FullName);
+            return hostBuilder.ConfigureServices(services => services.AddSingleton<IServerFactory, ServerFactory>());
+        }
+
+        public static IWebHostBuilder UseKestrel(this IWebHostBuilder hostBuilder, Action<KestrelServerOptions> options)
```

Comment:
Is there some special reason we need to configure the options via WebHostBuilder rather than doing it in Start.ConfigureServices with a `ConfigureKestrel(Action<KestrelServerOptions> options)` extension like we'd use for middleware? This won't execute until after ConfigureServices anyways. And then devs can use services injected via Startup's ctor.